CSRF update (#160)
OWASP/CheatSheetSeries: cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.md

### Synchronizer Token Pattern

> Any state changing operation requires a secure random token (e.g., CSRF token) to prevent CSRF attacks. A CSRF token should be unique per user session, large random value, and also generated by a cryptographically secure random number generator. The CSRF token is added as a hidden field for forms, headers/parameters for AJAX calls (If you need to add it to header, it is better to make sure that the token header is not being logged at your server) and within the URL if the state changing operation occurs via a GET. See "Disclosure of Token in URL" section below. The server rejects the requested action if the CSRF token fails validation.

Reviewer: Add a link to the section instead of the name "Disclosure of Token in URL"

Author: done :)

> Any state changing operation requires a secure random token (e.g., CSRF token) to prevent CSRF attacks. A CSRF token should be unique per user session, large random value, and also generated by a cryptographically secure random number generator. The CSRF token is added as a hidden field for forms, headers/parameters for AJAX calls (If you need to add it to header, it is better to make sure that the token header is not being logged at your server) and within the URL if the state changing operation occurs via a GET. See "Disclosure of Token in URL" section below. The server rejects the requested action if the CSRF token fails validation.

Reviewer: Check if you can handle the grammar fix of the PR #151 in your PR directly

Author: done!

> Any state changing operation requires a secure random token (e.g., CSRF token) to prevent CSRF attacks. A CSRF token should be unique per user session, a large random value, and also generated by a cryptographically secure random number generator. The CSRF token is added as a hidden field for forms, headers/parameters for AJAX calls (If you need to add it to header, it is better to make sure that the token header is not being logged at your server) and within the URL if the state changing operation occurs via a GET. See "Disclosure of Token in URL" section below. The server rejects the requested action if the CSRF token fails validation.

Reviewer 2: 1. Why are we saying that most practices implement CSRF in a hidden field? We can simply mention the techniques (in the headers, hidden fields, etc.) to do and where they can be used (forms, AJAX calls, etc.)
2. Instead of saying `See Disclosure of Token ...`, directly link to the section

Reviewer 3: @ThunderSon I am not sure if I got your first point. Where did we mention that "most practices implement CSRF in a hidden field"?? We mentioned "The CSRF token is added as a hidden field for forms, headers/parameters for AJAX calls"

Reviewer 2: Exactly. Who said that? We can simply mention that CSRF token can be added through hidden fields, headers, additional parameters in forms, AJAX calls, etc. Just mention all possible techniques to add the token for all state changing actions

Author: done

Reviewer 3: I have not seen many code samples where people include tokens in headers with form submissions and hence suggested to mention it considering the regular trend. Anyway, I am fine with the way its edited now.

> Any state changing operation requires a secure random token (e.g., CSRF token) to prevent CSRF attacks. A CSRF token should be unique per user session, a large random value, and also generated by a cryptographically secure random number generator. The CSRF token can be added through hidden fields, headers, additional parameters in forms, AJAX calls and within the URL. Token should be added to all state changing actions. Make sure that token will not leakue for example in server logs (e.g. when server log all of the headers) or [in the URL](#disclosure-of-token-in-url). The server rejects the requested action if the CSRF token fails validation.

Reviewer 2: What do you think about the following suggestion where I:
- Improved the way the CSRF token is described
- Improved the ease of readability
- Removed sending the CSRF token as part of the URL. It doesn't make sense sending it in the URL and stating right after that URLs get logged and we are saying not to log it?
- Made a small separation between where the CSRF token is added, and what functions use it.
```suggestion
Any state changing operation requires a secure random token (e.g., CSRF token) to prevent CSRF attacks. A CSRF token should be:

- unique per user session;
- large random value;
- generated by a Cryptographically Secure Pseudo-Random Number Generator (CSPRNG).

The CSRF token can be added through hidden fields, headers, and can be used with forms, and AJAX calls. That token should be added to all state changing actions. Make sure that the token is not leaked in the server logs, or [in the URL](#disclosure-of-token-in-url). The server must reject the requested action if the CSRF token fails validation.
```

Author: Looks much better! 
I see also one one change that we can add here. In the first sentence we write "Any state changing operation requires a secure random token" and that we repet it "That token should be added to all state changing actions". Maybe we should delete this repetition?

Reviewer 2: Certainly!

> **Disclosure of Token in URL**

Reviewer 2: Why is this not a subsection?

Author: it is now :)

The Strict value will prevent the cookie from being sent by the browser to the target site in all cross-site browsing context, even when following a regular link. For example, for a GitHub-like website this would mean that if a logged-in user follows a link to a private GitHub project posted on a corporate discussion forum or email, GitHub will not receive the session cookie and the user will not be able to access the project. A bank website however doesn't want to allow any transactional pages to be linked from external sites, so the Strict flag would be most appropriate.

The default Lax value provides a reasonable balance between security and usability for websites that want to maintain user's logged-in session after the user arrives from an external link. In the above GitHub scenario, the session cookie would be allowed when following a regular link from an external website while blocking it in CSRF-prone request methods such as POST. Only cross-site-requests that are allowed in Lax mode are the ones that have top-level navigations and are also [safe](https://tools.ietf.org/html/rfc7231#section-4.2.1) HTTP methods.

> For more details on the `SameSite` values, check the following [section](https://tools.ietf.org/html/draft-ietf-httpbis-rfc6265bis-02#section-5.3.7.1) from the [rfc](https://tools.ietf.org/html/draft-ietf-httpbis-rfc6265bis-02).

Reviewer: Add a section explaining what does the `None` value like for the 2 others ones.

Author: @ThunderSon you recently updated `SameSite` section. From what I remember there was a reason why `None` is not covered.
I know that `None` will be soon treated different in Chrome

Author: This PR is for for improving Samesite Cookie Attribute I just moved this paragraph without changes. You can open a new issue for it.

Reviewer 2: Exactly. This shouldn't be handled in this PR.

# Defense In Depth Techniques

> ## Samesite Cookie Attribute

Reviewer 3: Any specific reason on why we are moving this section ahead in "defense in depth" measures?

Author: Because it is the easiest one and IMO will add at least the same level of security as other techniques in this section. In addition to that this is true defense in depth because it protect against CSRF on other layer (browser instead of server side).

The default Lax value provides a reasonable balance between security and usability for websites that want to maintain user's logged-in session after the user arrives from an external link. In the above GitHub scenario, the session cookie would be allowed when following a regular link from an external website while blocking it in CSRF-prone request methods such as POST. Only cross-site-requests that are allowed in Lax mode are the ones that have top-level navigations and are also [safe](https://tools.ietf.org/html/rfc7231#section-4.2.1) HTTP methods.

For more details on the `SameSite` values, check the following [section](https://tools.ietf.org/html/draft-ietf-httpbis-rfc6265bis-02#section-5.3.7.1) from the [rfc](https://tools.ietf.org/html/draft-ietf-httpbis-rfc6265bis-02).

Example of cookies using this attribute:

```
Set-Cookie: JSESSIONID=xxxxx; SameSite=Strict
Set-Cookie: JSESSIONID=xxxxx; SameSite=Lax
```

All desktop browsers and almost all mobile browsers now support the `SameSite` attribute. To keep track of the browsers implementing it and the usage of the attribute, refer to the following [service](https://caniuse.com/#feat=same-site-cookie-attribute).

> It is important to note that this attribute should be implemented as an additional layer *defense in depth* concept. This attribute protects the user through the browsers supporting it, and it contains as well 2 ways to bypass it as mentioned in the following [section](https://tools.ietf.org/html/draft-ietf-httpbis-rfc6265bis-02#section-5.3.7.1). This attribute should not replace having a CSRF Token. Instead, it should co-exist with that token in order to protect the user in a more robust way.

Reviewer 3: In the earlier paragraph, we mention that all desktop browsers and almost all mobile browsers now support this attribute and here we say that "protects the user through the browsers supporting it". Its not very clear on what we are trying to explain. How about we re-frame the second section as below??

Even though most newer browser versions support this attribute, old versions do not in which case browsers just ignore this attribute sent by the server. Since we cannot rely on user browsers for providing security, this defense should only be implemented as an *defense in depth* concept. This defense also contains 2 ways to bypass it as mentioned in the following [section](https://tools.ietf.org/html/draft-ietf-httpbis-rfc6265bis-02#section-5.3.7.1). Hence, it should not replace having a CSRF Token and should co-exist with it in order to protect users in a more robust way.

Author: The same answer as I give to Dom - This PR is for for improving Samesite Cookie Attribute I just moved this paragraph without changes. You can open a new issue for it.

- There are some [privacy contexts](https://wiki.mozilla.org/Security/Origin#Privacy-Sensitive_Contexts) where Origin is set to “null” For example, see the following [here](https://www.google.com/search?q=origin+header+sent+null+value+site%3Astackoverflow.com&oq=origin+header+sent+null+value+site%3Astackoverflow.com).
- Origin header is included for all cross origin requests but for same origin requests, in most browsers it is only included in POST/DELETE/PUT **Note:** Although it is not ideal, many developers use GET requests to do state changing operations.
- Referer header is no exception. There are multiple use cases where referrer header is omitted as well ([1](https://stackoverflow.com/questions/6880659/in-what-cases-will-http-referer-be-empty), [2](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referer), [3](https://en.wikipedia.org/wiki/HTTP_referer#Referer_hiding), [4](https://seclab.stanford.edu/websec/csrf/csrf.pdf) and [5](https://www.google.com/search?q=referrer+header+sent+null+value+site:stackoverflow.com)). Load balancers, proxies and embedded network devices are also well known to strip the referrer header due to privacy reasons in logging them.

> 

Reviewer 3: I am fine removing the last three paragraphs but the first paragraph is what that makes this technique "usable" as even as "defense in depth".

Author: true, I will add it again

*b)   If an attacker is in the middle, they can usually force a request to the same domain over HTTP. If an application is hosted at `https://secure.example.com`, even if the cookies are set with the secure flag, a man in the middle can force connections to `http://secure.example.com` and set (overwrite) any arbitrary cookies (even though the secure flag prevents the attacker from reading those cookies). Even if the HSTS header is set on the server and the browser visiting the site supports HSTS (this would prevent a man in the middle from forcing plaintext HTTP requests) unless the HSTS header is set in a way that includes all subdomains, a man in the middle can simply force a request to a separate subdomain and overwrite cookies similar to 1. In other words, as long as `http://hellokitty.marketing.example.com` doesn't force https, then an attacker can overwrite cookies on any `example.com` subdomain.*"

> 

Reviewer 3: Not sure why this is removed. I added this workaround to mitigate the two issues mentioned in this technique above.

Author: Agree! I will add them again

- Origin header is included for all cross origin requests but for same origin requests, in most browsers it is only included in POST/DELETE/PUT **Note:** Although it is not ideal, many developers use GET requests to do state changing operations.
- Referer header is no exception. There are multiple use cases where referrer header is omitted as well ([1](https://stackoverflow.com/questions/6880659/in-what-cases-will-http-referer-be-empty), [2](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referer), [3](https://en.wikipedia.org/wiki/HTTP_referer#Referer_hiding), [4](https://seclab.stanford.edu/websec/csrf/csrf.pdf) and [5](https://www.google.com/search?q=referrer+header+sent+null+value+site:stackoverflow.com)). Load balancers, proxies and embedded network devices are also well known to strip the referrer header due to privacy reasons in logging them.

> Usually, a minor percentage of traffic does fall under above categories ([1-2%](http://homakov.blogspot.com/2012/04/playing-with-referer-origin-disquscom.html)) and no enterprise would want to lose even this minor % of traffic. One of the popular technique used across Internet to make this technique more usable is to accept the request if the Origin/referrer matches your configured list of domains "OR" a null value (Examples [here](http://homakov.blogspot.com/2012/04/playing-with-referer-origin-disquscom.html). null value is to cover the edge cases mentioned above where these headers are not sent). Please note that, attackers can use it to exploit CSRF using the edge cases but many prefer to deploy this technique it as a defense in depth measure because of minor effort involved in creating it.

Reviewer 3: Better rephrase? (I understand that I initially wrote it but on re-reading it, I felt it can be improved a bit)

_"Please note that, attackers can exploit this but people prefer to use this technique as a defense in depth measure because of minor effort involved in deploying it "_

Reviewer 3: Once this change is made and stripped version of "not so popular mitigation" techniques is added , I am fine with this pull request. Thanks again for this refactoring :)

This defense technique is specifically discussed in section 4.3 of [Robust Defenses for Cross-Site Request Forgery](https://seclab.stanford.edu/websec/csrf/csrf.pdf). However, bypasses of this defense using Flash were documented as early as 2008 and again as recently as 2015 by Mathias Karlsson to [exploit a CSRF flaw in Vimeo](https://hackerone.com/reports/44146). Riyaz Walikar from [Appsecco Labs](https://appsecco.com/) was able to exploit this even in early 2019 [using older versions of Firefox](https://blog.appsecco.com/exploiting-csrf-on-json-endpoints-with-flash-and-redirects-681d4ad6b31b) and also in some recent versions of Chrome (with a bit of [user involvement](https://github.com/sp1d3r/swf_json_csrf/blob/master/read.html)). Since we cannot the control the versions of browser end user uses nor depend on it for their security, this technique is not recommended as a primary defense in measure.

> This technique obviously works for Ajax calls and you have to still need protect `<form>` tags with approaches described in this document such as tokens. Setting custom headers on form tags is not possible directly. Also, CORS configuration should also be robust to make this solution work effectively (as custom headers for requests coming from other domains trigger a pre-flight CORS check).

Reviewer: *Setting custom headers on form tags is not possible directly*

Explains what are the possibilities to address that or add a ref to the section of the document talking about this.

Reviewer 3: I remember adding this statement long back. I think its safe to remove it.

Author: Ok so I will delete this

While these are a very strong CSRF defense, it does create a huge impact on the user experience. For applications that are in need of high security for some operations (password change, money transfer etc.), these techniques should be used along with token based mitigation. Please note that tokens by themselves can mitigate CSRF, developers should use these techniques only to achieve additional security for their high sensitive operations.

> 

Reviewer 3: I  personally know lot of developers who think content-type header validation can be used as a CSRF defense and know few folks who think triple submit cookie is better than double submit cookie.

My understanding of cheat sheet is that when people come to it, they need to know whats right way to do and whats wrong way. We dont need to highlight all possible "wrong" ways but at least mention the most widely known one's. Hence, I recommend adding these things back. We can add it with different title/ strip them to a line/two instead of how they are expanded now. What do you guys think?

Author: I agree, it would be awesome to leave them but in shorter from!

Reviewer 3: Yup, fine with me.
